channeldb/migrations.go

```diff
@@ -458,3 +459,105 @@ func paymentStatusesMigration(tx *bolt.Tx) error {
 
 	return nil
 }
+
+// migratePruneEdgeUpdateIndex is a database migration that attempts to resolve
+// some lingering bugs with regards to edge policies and their update index.
+// Stale entries within the edge update index were not being properly pruned due
+// to a miscalculation on the offset of an edge's policy last update. This
+// migration also fixes the case where the public keys within edge policies were
+// being serialized with an extra byte, causing an even greater error when
+// attempting to perform the offset calculation described earlier.
+func migratePruneEdgeUpdateIndex(tx *bolt.Tx) error {
+	// To begin the migration, we'll retrieve the update index bucket. If it
+	// does not exist, we have nothing left to do so we can simply exit.
+	edges := tx.Bucket(edgeBucket)
+	if edges == nil {
+		return nil
+	}
+	edgeUpdateIndex := edges.Bucket(edgeUpdateIndexBucket)
+	if edgeUpdateIndex == nil {
+		return nil
+	}
+
+	// Retrieve some buckets that will be needed later on. These should
+	// already exist given the assumption that the buckets above do as well.
+	edgeIndex := edges.Bucket(edgeIndexBucket)
+	if edgeIndex == nil {
+		return errors.New("edge index should exist but does not")
+	}
+	nodes := tx.Bucket(nodeBucket)
+	if nodes == nil {
+		return errors.New("node bucket should exist but does not")
+	}
+
+	log.Info("Migrating database to properly prune edge update index")
+
+	// We'll need to properly prune all the outdated entries within the edge
+	// update index. To do so, we'll gather all of the existing policies
+	// within the graph to re-populate them later on.
+	var edgeKeys [][]byte
+	err := edges.ForEach(func(edgeKey, edgePolicyBytes []byte) error {
+		// All valid entries are indexed by a public key (33 bytes)
+		// followed by a channel ID (8 bytes), so we'll skip any entries
+		// with keys that do not match this.
+		if len(edgeKey) != 33+8 {
+			return nil
+		}
+
+		// Skip any entries with unknown policies.
+		if bytes.Equal(edgePolicyBytes[:], unknownPolicy) {
```

Comment:
We want to keep the unknown policies, as they are needed for path finding.

Reply:
They aren't being deleted from the `edgeBucket`, so they will still exist. Edges with unknown policies will not exist within the update index, so it's fine to skip them.

Reply:
Ok, I see :) To make this more clear I think it would be better to just add all edge keys here, then when you get the edge policy below, check if is an unknown policy:

```go
for _, edgeKey := range edgeKeys {
		edgePolicyBytes := edges.Get(edgeKey)

                 // If the edge policy for this edgeKey (nodeID || chanID) is unknown, we don't have to re-add it to the 
                 // bucket. This is safe since edgeKey's with unknown policies are never added to the edge update index.
                 if bytes.Equal(edgePolicyBytes[:], unknownPolicy) {
                             continue
                 }
		edgePolicy, err := deserializeChanEdgePolicy(
```